File: apps/backend/kubedeck_backend/api/validation.py
from __future__ import annotations

import re

from fastapi import HTTPException

from kubedeck_backend.core.models import OperationConfirmation

MAX_LOG_TAIL_LINES = 5000
IDENTIFIER_PATTERN = re.compile(r"^[A-Za-z0-9._:-]+$")


def api_error(status_code: int, code: str, message: str, raw_stderr: str = "", command_preview: str = "") -> HTTPException:
    return HTTPException(
        status_code=status_code,
        detail={"code": code, "message": message, "rawStderr": raw_stderr, "commandPreview": command_preview},
    )


def validate_identifier(value: str, field: str, *, max_length: int = 253) -> str:
    text = (value or "").strip()
    if not text:
        raise api_error(400, "INVALID_IDENTIFIER", f"{field} must not be empty")
    if len(text) > max_length:
        raise api_error(400, "INVALID_IDENTIFIER", f"{field} is too long")
    if any(char in text for char in ("/", "\\", "\x00")):
        raise api_error(400, "INVALID_IDENTIFIER", f"{field} contains an invalid path separator")
    if not IDENTIFIER_PATTERN.match(text):
        raise api_error(400, "INVALID_IDENTIFIER", f"{field} contains unsupported characters")
    return text


def normalize_tail_lines(value: int) -> int:
    try:
        tail = int(value)
    except Exception:
        return 500
    return max(1, min(MAX_LOG_TAIL_LINES, tail))


def ensure_payload_size(text: str, max_bytes: int, label: str) -> None:
    size = len((text or "").encode("utf-8", "replace"))
    if size > max_bytes:
        raise api_error(413, "PAYLOAD_TOO_LARGE", f"{label} is too large ({size} bytes, limit {max_bytes} bytes)")


def require_confirmation(
    confirmation: OperationConfirmation | None,
    cluster_id: str,
    action: str,
    resource: str,
    namespace: str,
    name: str,
    expected_typed_name: str | None = None,
) -> None:
    if confirmation is None:
        raise HTTPException(status_code=400, detail={"code": "CONFIRMATION_REQUIRED", "message": f"Confirmation is required for {action}", "rawStderr": "", "commandPreview": ""})
    if confirmation.clusterId != cluster_id:
        raise HTTPException(status_code=400, detail={"code": "CONFIRMATION_CLUSTER_MISMATCH", "message": "Confirmation cluster does not match request", "rawStderr": "", "commandPreview": ""})
    if confirmation.action != action:
        raise HTTPException(status_code=400, detail={"code": "CONFIRMATION_ACTION_MISMATCH", "message": "Confirmation action does not match request", "rawStderr": "", "commandPreview": ""})
    if (confirmation.resource or resource) != resource:
        raise HTTPException(status_code=400, detail={"code": "CONFIRMATION_RESOURCE_MISMATCH", "message": "Confirmation resource does not match request", "rawStderr": "", "commandPreview": ""})
    if (confirmation.namespace or namespace) != namespace:
        raise HTTPException(status_code=400, detail={"code": "CONFIRMATION_NAMESPACE_MISMATCH", "message": "Confirmation namespace does not match request", "rawStderr": "", "commandPreview": ""})
    if (confirmation.name or name) != name:
        raise HTTPException(status_code=400, detail={"code": "CONFIRMATION_NAME_MISMATCH", "message": "Confirmation name does not match request", "rawStderr": "", "commandPreview": ""})
    if expected_typed_name is not None and confirmation.typedName != expected_typed_name:
        raise HTTPException(status_code=400, detail={"code": "CONFIRMATION_TYPED_NAME_MISMATCH", "message": "Typed confirmation value is invalid", "rawStderr": "", "commandPreview": ""})

File: apps/desktop/src/main/backend/validation.ts
export const MAX_LOG_TAIL_LINES = 5000;

const IDENTIFIER_PATTERN = /^[A-Za-z0-9._:-]+$/;

export class RequestValidationError extends Error {
  constructor(
    readonly statusCode: number,
    readonly code: string,
    message: string,
  ) {
    super(message);
  }
}

export function validateIdentifier(
  value: string,
  field: string,
  maxLength = 253,
): string {
  const text = String(value ?? "").trim();

  if (!text) {
    throw new RequestValidationError(400, "INVALID_IDENTIFIER", `${field} must not be empty`);
  }
  if (text.length > maxLength) {
    throw new RequestValidationError(400, "INVALID_IDENTIFIER", `${field} is too long`);
  }
  if (text.includes("/") || text.includes("\\") || text.includes("\0")) {
    throw new RequestValidationError(
      400,
      "INVALID_IDENTIFIER",
      `${field} contains an invalid path separator`,
    );
  }
  if (!IDENTIFIER_PATTERN.test(text)) {
    throw new RequestValidationError(
      400,
      "INVALID_IDENTIFIER",
      `${field} contains unsupported characters`,
    );
  }

  return text;
}

export function normalizeTailLines(value: string | null): number {
  if (value === null || value.trim() === "") return 500;
  if (!/^-?\d+$/.test(value.trim())) {
    throw new RequestValidationError(422, "INVALID_QUERY", "tail must be an integer");
  }

  const parsed = Number.parseInt(value, 10);
  return Math.max(1, Math.min(MAX_LOG_TAIL_LINES, parsed));
}

export function parseBooleanQuery(
  value: string | null,
  field: string,
  defaultValue = false,
): boolean {
  if (value === null || value.trim() === "") return defaultValue;

  const normalized = value.trim().toLowerCase();
  if (["true", "1", "yes", "on"].includes(normalized)) return true;
  if (["false", "0", "no", "off"].includes(normalized)) return false;

  throw new RequestValidationError(422, "INVALID_QUERY", `${field} must be a boolean`);
}
